File: crates/solid-linter/src/rules/imports.rs
//! solid/imports
//!
//! Enforce consistent imports from solid-js modules.
//! This rule detects imports from wrong sources and suggests the correct one.

use oxc_ast::ast::{ImportDeclaration, ImportDeclarationSpecifier, Program};
use oxc_ast_visit::{walk, Visit};
use oxc_span::GetSpan;

use crate::context::LintContext;
use crate::diagnostic::Diagnostic;
use crate::{RuleCategory, RuleMeta};

const SOLID_JS_PRIMITIVES: &[&str] = &[
    "createSignal",
    "createEffect",
    "createMemo",
    "createResource",
    "onMount",
    "onCleanup",
    "onError",
    "untrack",
    "batch",
    "on",
    "createRoot",
    "getOwner",
    "runWithOwner",
    "mergeProps",
    "splitProps",
    "useTransition",
    "observable",
    "from",
    "mapArray",
    "indexArray",
    "createContext",
    "useContext",
    "children",
    "lazy",
    "createUniqueId",
    "createDeferred",
    "createRenderEffect",
    "createComputed",
    "createReaction",
    "createSelector",
    "DEV",
    "For",
    "Show",
    "Switch",
    "Match",
    "Index",
    "ErrorBoundary",
    "Suspense",
    "SuspenseList",
];

const SOLID_JS_WEB: &[&str] = &[
    "Portal",
    "render",
    "hydrate",
    "renderToString",
    "renderToStream",
    "isServer",
    "renderToStringAsync",
    "generateHydrationScript",
    "HydrationScript",
    "Dynamic",
];

const SOLID_JS_STORE: &[&str] = &[
    "createStore",
    "produce",
    "reconcile",
    "unwrap",
    "createMutable",
    "modifyMutable",
];

#[derive(Debug, Clone, Default)]
pub struct Imports;

impl RuleMeta for Imports {
    const NAME: &'static str = "solid/imports";
    const CATEGORY: RuleCategory = RuleCategory::BestPractices;
}

impl Imports {
    pub fn new() -> Self {
        Self
    }

    pub fn check_program<'a>(
        &self,
        program: &Program<'a>,
        ctx: &LintContext<'a>,
    ) -> Vec<Diagnostic> {
        let mut visitor = ImportsVisitor::new(ctx);
        visitor.visit_program(program);
        visitor.diagnostics
    }
}

fn get_correct_source(name: &str) -> Option<&'static str> {
    if SOLID_JS_PRIMITIVES.contains(&name) {
        Some("solid-js")
    } else if SOLID_JS_WEB.contains(&name) {
        Some("solid-js/web")
    } else if SOLID_JS_STORE.contains(&name) {
        Some("solid-js/store")
    } else {
        None
    }
}

fn is_solid_source(source: &str) -> bool {
    source == "solid-js"
        || source == "solid-js/web"
        || source == "solid-js/store"
        || source.starts_with("solid-js/")
}

struct ImportsVisitor<'a, 'ctx> {
    #[allow(dead_code)]
    ctx: &'ctx LintContext<'a>,
    diagnostics: Vec<Diagnostic>,
}

impl<'a, 'ctx> ImportsVisitor<'a, 'ctx> {
    fn new(ctx: &'ctx LintContext<'a>) -> Self {
        Self {
            ctx,
            diagnostics: Vec::new(),
        }
    }

    fn check_import_declaration(&mut self, decl: &ImportDeclaration<'a>) {
        let source = decl.source.value.as_str();

        if !is_solid_source(source) {
            return;
        }

        if let Some(specifiers) = &decl.specifiers {
            for specifier in specifiers {
                if let ImportDeclarationSpecifier::ImportSpecifier(import_specifier) = specifier {
                    let imported_name = import_specifier.imported.name().as_str();

                    if let Some(correct_source) = get_correct_source(imported_name) {
                        if source != correct_source {
                            self.diagnostics.push(Diagnostic::warning(
                                Imports::NAME,
                                import_specifier.span(),
                                format!(
                                    "Prefer importing {} from \"{}\".",
                                    imported_name, correct_source
                                ),
                            ));
                        }
                    }
                }
            }
        }
    }
}

impl<'a, 'ctx> Visit<'a> for ImportsVisitor<'a, 'ctx> {
    fn visit_import_declaration(&mut self, decl: &ImportDeclaration<'a>) {
        self.check_import_declaration(decl);
        walk::walk_import_declaration(self, decl);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use oxc_allocator::Allocator;
    use oxc_parser::Parser;
    use oxc_span::SourceType;

    fn check_code(source: &str) -> Vec<Diagnostic> {
        let allocator = Allocator::default();
        let source_type = SourceType::default().with_module(true).with_jsx(true);

        let parser = Parser::new(&allocator, source, source_type);
        let parsed = parser.parse();

        if !parsed.errors.is_empty() {
            panic!("Parse errors: {:?}", parsed.errors);
        }

        let ctx = LintContext::new(source, source_type);
        let rule = Imports::new();
        rule.check_program(&parsed.program, &ctx)
    }

    #[test]
    fn test_rule_name() {
        assert_eq!(Imports::NAME, "solid/imports");
    }

    #[test]
    fn test_rule_category() {
        assert_eq!(Imports::CATEGORY, RuleCategory::BestPractices);
    }

    // ===== Valid cases =====

    #[test]
    fn test_correct_solid_js_import_valid() {
        let diagnostics = check_code(r#"import { createSignal } from "solid-js";"#);
        assert!(
            diagnostics.is_empty(),
            "Expected no diagnostics but got: {:?}",
            diagnostics
        );
    }

    #[test]
    fn test_correct_solid_js_web_import_valid() {
        let diagnostics = check_code(r#"import { render } from "solid-js/web";"#);
        assert!(
            diagnostics.is_empty(),
            "Expected no diagnostics but got: {:?}",
            diagnostics
        );
    }

    #[test]
    fn test_correct_solid_js_store_import_valid() {
        let diagnostics = check_code(r#"import { createStore } from "solid-js/store";"#);
        assert!(
            diagnostics.is_empty(),
            "Expected no diagnostics but got: {:?}",
            diagnostics
        );
    }

    #[test]
    fn test_multiple_correct_imports_valid() {
        let diagnostics =
            check_code(r#"import { createSignal, createEffect, createMemo } from "solid-js";"#);
        assert!(
            diagnostics.is_empty(),
            "Expected no diagnostics but got: {:?}",
            diagnostics
        );
    }

    #[test]
    fn test_non_solid_import_valid() {
        let diagnostics = check_code(r#"import { useState } from "react";"#);
        assert!(
            diagnostics.is_empty(),
            "Expected no diagnostics but got: {:?}",
            diagnostics
        );
    }

    #[test]
    fn test_unknown_solid_export_valid() {
        let diagnostics = check_code(r#"import { unknownExport } from "solid-js";"#);
        assert!(
            diagnostics.is_empty(),
            "Expected no diagnostics but got: {:?}",
            diagnostics
        );
    }

    // ===== Invalid cases =====

    #[test]
    fn test_render_from_solid_js_invalid() {
        let diagnostics = check_code(r#"import { render } from "solid-js";"#);
        assert!(
            !diagnostics.is_empty(),
            "Expected diagnostics for render from solid-js"
        );
        assert!(diagnostics[0].message.contains("solid-js/web"));
    }

    #[test]
    fn test_create_store_from_solid_js_invalid() {
        let diagnostics = check_code(r#"import { createStore } from "solid-js";"#);
        assert!(
            !diagnostics.is_empty(),
            "Expected diagnostics for createStore from solid-js"
        );
        assert!(diagnostics[0].message.contains("solid-js/store"));
    }

    #[test]
    fn test_create_signal_from_solid_js_web_invalid() {
        let diagnostics = check_code(r#"import { createSignal } from "solid-js/web";"#);
        assert!(
            !diagnostics.is_empty(),
            "Expected diagnostics for createSignal from solid-js/web"
        );
        assert!(diagnostics[0].message.contains("solid-js"));
    }

    #[test]
    fn test_portal_from_solid_js_invalid() {
        let diagnostics = check_code(r#"import { Portal } from "solid-js";"#);
        assert!(
            !diagnostics.is_empty(),
            "Expected diagnostics for Portal from solid-js"
        );
        assert!(diagnostics[0].message.contains("solid-js/web"));
    }

    #[test]
    fn test_mixed_correct_and_incorrect_imports_invalid() {
        let diagnostics =
            check_code(r#"import { createSignal, render } from "solid-js";"#);
        assert_eq!(
            diagnostics.len(),
            1,
            "Expected 1 diagnostic for render"
        );
        assert!(diagnostics[0].message.contains("render"));
        assert!(diagnostics[0].message.contains("solid-js/web"));
    }

    #[test]
    fn test_produce_from_solid_js_invalid() {
        let diagnostics = check_code(r#"import { produce } from "solid-js";"#);
        assert!(
            !diagnostics.is_empty(),
            "Expected diagnostics for produce from solid-js"
        );
        assert!(diagnostics[0].message.contains("solid-js/store"));
    }

    #[test]
    fn test_is_server_from_solid_js_invalid() {
        let diagnostics = check_code(r#"import { isServer } from "solid-js";"#);
        assert!(
            !diagnostics.is_empty(),
            "Expected diagnostics for isServer from solid-js"
        );
        assert!(diagnostics[0].message.contains("solid-js/web"));
    }
}

File: packages/eslint-plugin-solid/src/rules/imports.ts
import { TSESTree as T, TSESLint, ESLintUtils } from "@typescript-eslint/utils";
import { appendImports, insertImports, isSolidV2, removeSpecifier } from "../utils";
import { getSourceCode } from "../compat";

const createRule = ESLintUtils.RuleCreator.withoutDocs;

// Below: create maps of imports and types to designated import source.
// We could mess with `Object.keys(require("solid-js"))` to generate this, but requiring it from
// node activates the "node" export condition, which doesn't necessarily match what users will
// receive i.e. through bundlers. Instead, we're manually listing all of the public exports that
// should be imported from "solid-js", etc.
// ==============

type Source = "solid-js" | "solid-js/web" | "solid-js/store";

// Set up map of imports to module
const primitiveMap = new Map<string, Source>();
for (const primitive of [
  "createSignal",
  "createEffect",
  "createMemo",
  "createResource",
  "onMount",
  "onCleanup",
  "onError",
  "untrack",
  "batch",
  "on",
  "createRoot",
  "getOwner",
  "runWithOwner",
  "mergeProps",
  "splitProps",
  "useTransition",
  "observable",
  "from",
  "mapArray",
  "indexArray",
  "createContext",
  "useContext",
  "children",
  "lazy",
  "createUniqueId",
  "createDeferred",
  "createRenderEffect",
  "createComputed",
  "createReaction",
  "createSelector",
  "DEV",
  "For",
  "Show",
  "Switch",
  "Match",
  "Index",
  "ErrorBoundary",
  "Suspense",
  "SuspenseList",
]) {
  primitiveMap.set(primitive, "solid-js");
}
for (const primitive of [
  "Portal",
  "render",
  "hydrate",
  "renderToString",
  "renderToStream",
  "isServer",
  "renderToStringAsync",
  "generateHydrationScript",
  "HydrationScript",
  "Dynamic",
]) {
  primitiveMap.set(primitive, "solid-js/web");
}
for (const primitive of [
  "createStore",
  "produce",
  "reconcile",
  "unwrap",
  "createMutable",
  "modifyMutable",
]) {
  primitiveMap.set(primitive, "solid-js/store");
}

// Set up map of type imports to module
const typeMap = new Map<string, Source>();
for (const type of [
  "Signal",
  "Accessor",
  "Setter",
  "Resource",
  "ResourceActions",
  "ResourceOptions",
  "ResourceReturn",
  "ResourceFetcher",
  "InitializedResourceReturn",
  "Component",
  "VoidProps",
  "VoidComponent",
  "ParentProps",
  "ParentComponent",
  "FlowProps",
  "FlowComponent",
  "ValidComponent",
  "ComponentProps",
  "Ref",
  "MergeProps",
  "SplitPrips",
  "Context",
  "JSX",
  "ResolvedChildren",
  "MatchProps",
]) {
  typeMap.set(type, "solid-js");
}
for (const type of [/* "JSX", */ "MountableElement"]) {
  typeMap.set(type, "solid-js/web");
}
for (const type of ["StoreNode", "Store", "SetStoreFunction"]) {
  typeMap.set(type, "solid-js/store");
}

// Solid 2.0 moved these exports into core "solid-js" (and removed the old locations in
// some cases). Don't flag them when imported from the source listed here, so the rule
// gives correct guidance for both Solid 1.x and 2.0 projects.
const alternateSourceMap = new Map<string, Source>([
  ["createStore", "solid-js"],
  ["reconcile", "solid-js"],
  ["Store", "solid-js"],
  ["StoreNode", "solid-js"],
  ["SetStoreFunction", "solid-js"],
]);

// ==============
// Solid 2.0: "solid-js/store" no longer exists (store exports live in core),
// and "solid-js/web" became the "@solidjs/web" package. When `settings.solid.
// version` is 2, the rule *requires* the 2.0 locations. Lists verified against
// the RC source (packages/solid/src/index.ts, packages/solid-web/src/index.ts).
type SourceV2 = "solid-js" | "@solidjs/web";

const primitiveMapV2 = new Map<string, SourceV2>();
for (const primitive of [
  // reactive primitives
  "createSignal",
  "createMemo",
  "createStore",
  "createEffect",
  "createRenderEffect",
  "createProjection",
  "createOptimistic",
  "createOptimisticStore",
  "createReaction",
  "createRoot",
  "createOwner",
  "createContext",
  "useContext",
  "children",
  "lazy",
  "createUniqueId",
  "createErrorBoundary",
  "createLoadingBoundary",
  "createRevealOrder",
  // lifecycle and utilities
  "onCleanup",
  "onSettled",
  "untrack",
  "flush",
  "merge",
  "omit",
  "mapArray",
  "repeat",
  "getOwner",
  "getObserver",
  "runWithOwner",
  "isEqual",
  "isPending",
  "latest",
  "action",
  "reconcile",
  "snapshot",
  "resolve",
  "refresh",
  "flatten",
  "deep",
  "affects",
  "storePath",
  "DEV",
  // control flow components
  "For",
  "Repeat",
  "Show",
  "Switch",
  "Match",
  "Errored",
  "Loading",
  "Reveal",
]) {
  primitiveMapV2.set(primitive, "solid-js");
}
for (const primitive of [
  "Portal",
  "Dynamic",
  "render",
  "hydrate",
  "isServer",
  "renderToString",
  "renderToStream",
  "renderToStringAsync",
  "generateHydrationScript",
  "HydrationScript",
]) {
  primitiveMapV2.set(primitive, "@solidjs/web");
}

const typeMapV2 = new Map<string, SourceV2>();
for (const type of [
  "Signal",
  "Accessor",
  "Setter",
  "Component",
  "Context",
  "ResolvedChildren",
  "Store",
  "StoreNode",
  "StoreSetter",
  "StoreOptions",
  "SignalOptions",
  "MemoOptions",
  "EffectOptions",
  "Owner",
]) {
  typeMapV2.set(type, "solid-js");
}
typeMapV2.set("JSX", "@solidjs/web");

const sourceRegex = /^solid-js(?:\/web|\/store)?$/;
const isSource = (source: string): source is Source => sourceRegex.test(source);

// The legacy "solid-js/store" and "solid-js/web" sources are deliberately
// excluded here: in v2 mode those modules no longer exist and solid/removed-api
// owns reporting/rewriting them, so this rule only polices specifier placement
// between the surviving modules.
const sourceRegexV2 = /^(?:solid-js|@solidjs\/web)$/;
const isSourceV2 = (source: string): source is Source | "@solidjs/web" =>
  sourceRegexV2.test(source);

export default createRule({
  meta: {
    type: "suggestion",
    docs: {
      description:
        'Enforce consistent imports from "solid-js", "solid-js/web", and "solid-js/store".',
      url: "https://github.com/solidjs-community/eslint-plugin-solid/blob/main/packages/eslint-plugin-solid/docs/imports.md",
    },
    fixable: "code",
    schema: [],
    messages: {
      "prefer-source": 'Prefer importing {{name}} from "{{source}}".',
    },
  },
  defaultOptions: [],
  create(context) {
    const v2 = isSolidV2(context);
    return {
      ImportDeclaration(node) {
        const source = node.source.value;
        if (v2 ? !isSourceV2(source) : !isSource(source)) return;

        for (const specifier of node.specifiers) {
          if (specifier.type === "ImportSpecifier") {
            const isType = specifier.importKind === "type" || node.importKind === "type";
            const map = v2
              ? isType
                ? typeMapV2
                : primitiveMapV2
              : isType
              ? typeMap
              : primitiveMap;
            const importedName =
              specifier.imported.type === "Identifier"
                ? specifier.imported.name
                : specifier.imported.value;
            const correctSource = map.get(importedName);
            if (
              correctSource != null &&
              correctSource !== source &&
              (v2 || alternateSourceMap.get(importedName) !== source)
            ) {
              context.report({
                node: specifier,
                messageId: "prefer-source",
                data: {
                  name: importedName,
                  source: correctSource,
                },
                fix(fixer) {
                  const sourceCode = getSourceCode(context);
                  const program: T.Program = sourceCode.ast;
                  const correctDeclaration = program.body.find(
                    (node) =>
                      node.type === "ImportDeclaration" && node.source.value === correctSource
                  ) as T.ImportDeclaration | undefined;

                  if (correctDeclaration) {
                    return [
                      removeSpecifier(fixer, sourceCode, specifier),
                      appendImports(fixer, sourceCode, correctDeclaration, [
                        sourceCode.getText(specifier),
                      ]),
                    ].filter(Boolean) as Array<TSESLint.RuleFix>;
                  }

                  const firstSolidDeclaration = program.body.find(
                    (node) =>
                      node.type === "ImportDeclaration" &&
                      (v2 ? isSourceV2(node.source.value) : isSource(node.source.value))
                  ) as T.ImportDeclaration | undefined;
                  return [
                    removeSpecifier(fixer, sourceCode, specifier),
                    insertImports(
                      fixer,
                      sourceCode,
                      correctSource,
                      [sourceCode.getText(specifier)],
                      firstSolidDeclaration,
                      isType
                    ),
                  ];
                },
              });
            }
          }
        }
      },
    };
  },
});
